storage/innobase/log/log0online.cc

```diff
@@ -44,37 +45,392 @@ Online database log parsing for changed page tracking */
 #include "my_sys.h"   /* DEBUG_SYNC_C */
 #endif
 
+/** For the given minilog record type determine if the record has (space; page)
+associated with it.
+@param[in] type the minilog record type
+@return true if the record has (space; page) in it */
+static constexpr bool log_online_rec_has_page(mlog_id_t type) noexcept {
+  static_assert(MLOG_BIGGEST_TYPE == 65,
+                "New MTR types must be reviewed for page presence");
+  return type != MLOG_MULTI_REC_END && type != MLOG_DUMMY_RECORD &&
+         type != MLOG_COMP_PAGE_CREATE_SDI && type != MLOG_PAGE_CREATE_SDI &&
+         type != MLOG_TABLE_DYNAMIC_META
+#ifdef UNIV_LOG_LSN_DEBUG
+         && type != MLOG_LSN
+#endif
+      ;
+}
+
+/* Mutex protecting log_bmp_sys and log buffers there */
+static ib_mutex_t log_bmp_sys_mutex;
+
+/** A redo log byte buffer with associated LSN values */
+template <std::size_t CAPACITY>
+class log_buffer {
+ protected:
+  using buffer_type = std::array<byte, CAPACITY>;
+
+ public:
+  using size_type = typename buffer_type::size_type;
+  using const_iterator = typename buffer_type::const_iterator;
+
+ protected:
+  buffer_type buffer;
+  static const constexpr auto capacity = CAPACITY;
+  size_type current_size{0};
+  lsn_t start_lsn{0};
+  lsn_t current_lsn{0};
+  const_iterator current_ptr{buffer.cbegin()};
+  lsn_t limit_lsn{0};
+
+  log_buffer() noexcept {}
+
+#ifdef UNIV_DEBUG
+  MY_NODISCARD bool invariants() const noexcept {
+    ut_ad(mutex_own(&log_bmp_sys_mutex));
+    ut_ad(start_lsn <= current_lsn);
+    ut_ad(current_lsn <= limit_lsn);
+    ut_ad(static_cast<decltype(start_lsn)>(current_ptr - buffer.cbegin()) <=
+          current_size);
+    return true;
+  }
+#endif
+
+ public:
+  MY_NODISCARD const_iterator ccurrent() const noexcept {
+    ut_ad(mutex_own(&log_bmp_sys_mutex));
+    return current_ptr;
+  }
+
+  void set_limit(lsn_t limit) {
+    ut_ad(mutex_own(&log_bmp_sys_mutex));
+    ut_ad(limit >= limit_lsn);
+    limit_lsn = limit;
+  }
+
+  MY_NODISCARD lsn_t get_current_lsn() const noexcept {
+    ut_ad(mutex_own(&log_bmp_sys_mutex));
+    return current_lsn;
+  }
+};
+
 static const constexpr auto FOLLOW_SCAN_SIZE = 4 * UNIV_PAGE_SIZE_MAX;
 
+static_assert(FOLLOW_SCAN_SIZE % OS_FILE_LOG_BLOCK_SIZE == 0,
+              "FOLLOW_SCAN_SIZE must be a multiple of OS_FILE_LOG_BLOCK_SIZE");
+
+static const constexpr auto LOG_BLOCK_SIZE_NO_TRL =
+    OS_FILE_LOG_BLOCK_SIZE - LOG_BLOCK_TRL_SIZE;
+
+static const constexpr auto LOG_BLOCK_BOUNDARY_LSN_PAD =
+    LOG_BLOCK_HDR_SIZE + LOG_BLOCK_TRL_SIZE;
+
+/** The buffer for reading log data, filled in by recv_read_log_recs and moved
+in chunks to the parse buffer while skipping log block headers and trailers. */
+class log_read_buffer final : public log_buffer<FOLLOW_SCAN_SIZE> {
```

Comment:
Private inheritance as `log_buffer<FOLLOW_SCAN_SIZE>` is an implementation detail
```c++
private:
  using base_type = log_buffer<FOLLOW_SCAN_SIZE>;
public:
  using base_type::ccurrent;
  using base_type::set_limit;
  using base_type::get_current_lsn;
```

Using `log_buffer<FOLLOW_SCAN_SIZE>* ptr = new log_read_buffer()` should not be allowed (there is no virtual destructor and it is not needed).

Reply:
Actually no, both child class objects get their parent public methods called: ccurrent(), set_limit(), get_current_lsn()

Reply:
My main concern is still the same but theoretical. In the following usage scenario
```c++
log_buffer<FOLLOW_SCAN_SIZE>* ptr = new log_read_buffer();
delete ptr;
```
only the desctuctor of the base class will be called and therefore we will get memory leaks.
On the other hand, it has nothing in common with how these classes are used in your code. So, up to you.